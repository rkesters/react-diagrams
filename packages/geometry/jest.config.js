const sharedConfig = require('../../jest.config.js')(__dirname);
module.exports = {
	...sharedConfig,
	testEnvironment: 'jsdom',

};
