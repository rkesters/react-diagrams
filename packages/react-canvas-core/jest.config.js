const sharedConfig = require('../../jest.config.js')(__dirname);
module.exports = {
	...sharedConfig,
	testEnvironment: 'jsdom',
	collectCoverageFrom: [
		...sharedConfig.collectCoverageFrom,
		'!src/index.ts'
	],
};
