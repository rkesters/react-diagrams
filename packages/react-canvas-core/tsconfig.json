{
	"extends": "../../tsconfig.base.json",
	"compilerOptions": {
		"outDir": "dist",
		"rootDir": "src",
		"declarationDir": "dist/@types",
		"esModuleInterop": true
	},
	"include": [
		"./src"
	],
	"references": [
		{
			"path": "../geometry"
		}
	]
}
