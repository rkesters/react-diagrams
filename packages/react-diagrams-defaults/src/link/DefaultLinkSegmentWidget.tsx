import * as React from 'react';
import { DefaultLinkFactory } from './DefaultLinkFactory';
import { DiagramEngine } from '@projectstorm/react-diagrams-core';
import { DefaultLinkModel } from './DefaultLinkModel';

export interface DefaultLinkSegmentWidgetProps {
	path: string;
	link: DefaultLinkModel | null;
	selected: boolean;
	forwardRef: React.RefObject<SVGPathElement>;
	factory: DefaultLinkFactory | null | undefined;
	diagramEngine: DiagramEngine | null;
	onSelection: (selected: boolean) => any;
	extras: object;
}

export class DefaultLinkSegmentWidget extends React.Component<DefaultLinkSegmentWidgetProps> {
	render() {
		const factory = this.props.factory;
		const link = this.props.link;
		if (!factory || !link) {return (<> </>)}
		const Bottom = React.cloneElement(
			factory.generateLinkSegment(
				link,
				this.props.selected || link.isSelected(),
				this.props.path
			),
			{
				ref: this.props.forwardRef
			}
		);

		const Top = React.cloneElement(Bottom, {
			strokeLinecap: 'round',
			onMouseLeave: () => {
				this.props.onSelection(false);
			},
			onMouseEnter: () => {
				this.props.onSelection(true);
			},
			...this.props.extras,
			ref: null,
			'data-linkid': link.getID(),
			strokeOpacity: this.props.selected ? 0.1 : 0,
			strokeWidth: 20,
			fill: 'none',
			onContextMenu: () => {
				if (!link.isLocked()) {
					event?.preventDefault();
					link.remove();
				}
			}
		});

		return (
			<g>
				{Bottom}
				{Top}
			</g>
		);
	}
}
