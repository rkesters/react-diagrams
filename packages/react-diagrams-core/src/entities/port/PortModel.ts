import { NodeModel } from '../node/NodeModel';
import { LinkModel } from '../link/LinkModel';
import * as _ from 'lodash';
import { Point, Rectangle } from '@projectstorm/geometry';
import {
	BaseEntityEvent,
	BaseModelOptions,
	BasePositionModel,
	BasePositionModelGenerics,
	BasePositionModelListener,
	DeserializeEvent
} from '@projectstorm/react-canvas-core';

export enum PortModelAlignment {
	TOP = 'top',
	LEFT = 'left',
	BOTTOM = 'bottom',
	RIGHT = 'right'
}

export interface PortModelListener extends BasePositionModelListener {
	/**
	 * fires when it first receives positional information
	 */
	reportInitialPosition: (event: BaseEntityEvent<PortModel>) => void;
}

export interface PortModelOptions extends BaseModelOptions {
	alignment?: PortModelAlignment;
	maximumLinks?: number;
	name: string;
}

export interface PortModelGenerics extends BasePositionModelGenerics {
	OPTIONS: PortModelOptions;
	PARENT: NodeModel;
	LISTENER: PortModelListener;
}

export class PortModel<G extends PortModelGenerics = PortModelGenerics> extends BasePositionModel<G> {
	links: { [id: string]: LinkModel };

	// calculated post rendering so routing can be done correctly
	width: number;
	height: number;
	reportedPosition: boolean;

	constructor(options: G['OPTIONS']) {
		super(options);
		this.links = {};
		this.reportedPosition = false;
		this.width = 0;
		this.height = 0;
	}

	deserialize(event: DeserializeEvent<ReturnType<PortModel['serialize']>>) {
		super.deserialize(event);
		this.reportedPosition = false;
		this.options.name = event.data.name;
		this.options.alignment = event.data.alignment;
	}

	serialize() {
		return {
			...super.serialize(),
			name: this.options.name,
			alignment: this.options.alignment,
			parentNode: this.parent?.getID(),
			links: _.map(this.links, (link) => {
				return link.getID();
			})
		};
	}

	setPosition(point: Point): void;
	setPosition(x: number, y: number): void;
	setPosition(x: Point | number, y?: number) {
		let old = this.position;
		Point.isa(x) ? super.setPosition(x) : super.setPosition(x, y as number) ;
		const X = this.getPosition().x;
		const Y = this.getPosition().y;

		_.forEach(this.getLinks(), (link) => {
			let point = link.getPointForPort(this);
			point?.setPosition(point.getX() + X - old.x, point.getY() + Y - old.y);
		});
	}

	doClone(lookupTable = {}, clone: this) {
		clone.links = {};
		clone.parent = this.getParent()?.clone(lookupTable);
	}

	getNode(): NodeModel | undefined {
		return this.getParent();
	}

	getName(): string {
		return this.options.name;
	}

	getMaximumLinks(): number | undefined  {
		return this.options.maximumLinks;
	}

	setMaximumLinks(maximumLinks: number) {
		this.options.maximumLinks = maximumLinks;
	}

	removeLink(link: LinkModel) {
		delete this.links[link.getID()];
	}

	addLink(link: LinkModel) {
		this.links[link.getID()] = link;
	}

	getLinks(): { [id: string]: LinkModel } {
		return this.links;
	}

	public createLinkModel(): LinkModel | null {
		if (_.isFinite(this.options.maximumLinks)) {
			var numberOfLinks: number = _.size(this.links);
			if (this.options.maximumLinks === 1 && numberOfLinks >= 1) {
				return _.values(this.links)[0];
			} else if (numberOfLinks >= (this.options.maximumLinks ?? Number.MAX_SAFE_INTEGER/2)) {
				return null;
			}
		}
		return null;
	}

	reportPosition() {
		_.forEach(this.getLinks(), (link) => {
			link.getPointForPort(this)?.setPosition(this.getCenter());
		});
		this.fireEvent(
			{
				entity: this
			},
			'reportInitialPosition'
		);
	}

	getCenter(): Point {
		return new Point(this.getX() + this.width / 2, this.getY() + this.height / 2);
	}

	getBoundingBox(): Rectangle {
		return new Rectangle(this.position, this.width, this.height);
	}

	updateCoords(coords: Rectangle) {
		this.width = coords.getWidth();
		this.height = coords.getHeight();
		this.setPosition(coords.getTopLeft());
		this.reportedPosition = true;
		this.reportPosition();
	}

	canLinkToPort(port: PortModel): boolean {
		return true;
	}

	isLocked(): boolean {
		return super.isLocked() || (this.getParent()?.isLocked() ?? false);
	}
}
