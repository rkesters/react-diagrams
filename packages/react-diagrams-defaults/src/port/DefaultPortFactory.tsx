import { DefaultPortModel } from './DefaultPortModel';
import { AbstractModelFactory } from '@projectstorm/react-canvas-core';
import { DiagramEngine } from '@projectstorm/react-diagrams-core';

export class DefaultPortFactory extends AbstractModelFactory<DefaultPortModel, DiagramEngine> {
	constructor(engine: DiagramEngine) {
		super('default', engine);
	}

	generateModel(): DefaultPortModel {
		return new DefaultPortModel({
			name: 'unknown'
		});
	}
}
