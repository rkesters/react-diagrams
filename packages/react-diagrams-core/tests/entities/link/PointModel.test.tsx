import { Point } from '@projectstorm/geometry';
import { LinkModel } from '../../../src/entities/link/LinkModel';
import { PointModel, PointModelOptions } from '../../../src/entities/link/PointModel';
import { Toolkit } from '@projectstorm/react-canvas-core';
import { PortModel } from '../../../src/entities/port/PortModel';

describe('PointModel', () => {
	Toolkit.TESTING = true;
	Toolkit.TESTING_UID = 0;
	beforeEach(() => {
		Toolkit.TESTING = true;
		Toolkit.TESTING_UID = 0;
	});
	afterEach(() => {
		Toolkit.TESTING_UID = 0;
		Toolkit.TESTING = false;
	});
	test.each([
		{
			name: 'Minimal config',
			config: {
				link: new LinkModel({}),
				position: new Point(0, 0)
			},
			isConnectedToPort: false,
			isLocked: false
		},
		{
			name: 'Full config',
			config: {
				selected: false,
				extra: { foo: 'bar' },
				locked: false,
				link: new LinkModel({}),
				position: new Point(0, 0)
			},
			isConnectedToPort: false,
			isLocked: false
		},
		{
			name: 'Locked Parent',
			config: {
				link: new LinkModel({ locked: true }),
				position: new Point(0, 0)
			},
			isConnectedToPort: false,
			isLocked: true
		},
		{
			name: 'Locked',
			config: {
				link: new LinkModel({}),
				position: new Point(0, 0),
				locked: true
			},
			isConnectedToPort: false,
			isLocked: true
		},
		{
			name: 'Selected',
			config: {
				selected: true,
				link: new LinkModel({}),
				position: new Point(0, 0),
				locked: true
			},
			isConnectedToPort: false,
			isLocked: true
		},
		,
		{
			name: 'Connected',
			config: {
				selected: false,
				link: new LinkModel({}),
				position: new Point(0, 0),
				locked: false
			},
			isConnectedToPort: true,
			isLocked: false
		}
	])('All: $name', (tc) => {
		const { config, isConnectedToPort, isLocked } = tc;
		const model = new PointModel(config);

		if (isConnectedToPort) {
			const port = new PortModel({ name: 'source' });
			config.link.setPoints([model]);
			model.getParent().setSourcePort(port);
		}
		expect(model).toMatchSnapshot();
		expect(model.getLink()).toBe(config.link);
		expect(model.getParent()).toBe(config.link);
		expect(model.isConnectedToPort()).toBe(isConnectedToPort);
		expect(model.isLocked()).toBe(isLocked);

		isConnectedToPort
			? expect(model.getParent().getPoints()).toContain(model)
			: expect(model.getParent().getPoints()).not.toContain(model);

		model.remove();
		expect(model.getParent().getPoints()).not.toContain(model);
		model.setParent(null);
		model.remove();
	});
});
