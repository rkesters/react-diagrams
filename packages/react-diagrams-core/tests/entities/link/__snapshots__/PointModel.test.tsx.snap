// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointModel All: Connected 1`] = `
PointModel {
  "listeners": Object {},
  "options": Object {
    "id": "1",
    "link": LinkModel {
      "labels": Array [],
      "listeners": Object {},
      "options": Object {
        "id": "16",
      },
      "points": Array [
        [Circular],
      ],
      "renderedPaths": Array [],
      "sourcePort": PortModel {
        "links": Object {
          "16": [Circular],
        },
        "listeners": Object {},
        "options": Object {
          "id": "2",
          "name": "source",
        },
        "position": Point {
          "x": 0,
          "y": 0,
        },
        "reportedPosition": false,
      },
      "targetPort": null,
    },
    "locked": false,
    "position": Point {
      "x": 0,
      "y": 0,
    },
    "selected": false,
    "type": "point",
  },
  "parent": LinkModel {
    "labels": Array [],
    "listeners": Object {},
    "options": Object {
      "id": "16",
    },
    "points": Array [
      [Circular],
    ],
    "renderedPaths": Array [],
    "sourcePort": PortModel {
      "links": Object {
        "16": [Circular],
      },
      "listeners": Object {},
      "options": Object {
        "id": "2",
        "name": "source",
      },
      "position": Point {
        "x": 0,
        "y": 0,
      },
      "reportedPosition": false,
    },
    "targetPort": null,
  },
  "position": Point {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`PointModel All: Full config 1`] = `
PointModel {
  "listeners": Object {},
  "options": Object {
    "extra": Object {
      "foo": "bar",
    },
    "id": "1",
    "link": LinkModel {
      "labels": Array [],
      "listeners": Object {},
      "options": Object {
        "id": "4",
      },
      "points": Array [
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "5",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "6",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
      ],
      "renderedPaths": Array [],
      "sourcePort": null,
      "targetPort": null,
    },
    "locked": false,
    "position": Point {
      "x": 0,
      "y": 0,
    },
    "selected": false,
    "type": "point",
  },
  "parent": LinkModel {
    "labels": Array [],
    "listeners": Object {},
    "options": Object {
      "id": "4",
    },
    "points": Array [
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "5",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "6",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
    ],
    "renderedPaths": Array [],
    "sourcePort": null,
    "targetPort": null,
  },
  "position": Point {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`PointModel All: Locked 1`] = `
PointModel {
  "listeners": Object {},
  "options": Object {
    "id": "1",
    "link": LinkModel {
      "labels": Array [],
      "listeners": Object {},
      "options": Object {
        "id": "10",
      },
      "points": Array [
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "11",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "12",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
      ],
      "renderedPaths": Array [],
      "sourcePort": null,
      "targetPort": null,
    },
    "locked": true,
    "position": Point {
      "x": 0,
      "y": 0,
    },
    "type": "point",
  },
  "parent": LinkModel {
    "labels": Array [],
    "listeners": Object {},
    "options": Object {
      "id": "10",
    },
    "points": Array [
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "11",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "12",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
    ],
    "renderedPaths": Array [],
    "sourcePort": null,
    "targetPort": null,
  },
  "position": Point {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`PointModel All: Locked Parent 1`] = `
PointModel {
  "listeners": Object {},
  "options": Object {
    "id": "1",
    "link": LinkModel {
      "labels": Array [],
      "listeners": Object {},
      "options": Object {
        "id": "7",
        "locked": true,
      },
      "points": Array [
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "8",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "9",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
      ],
      "renderedPaths": Array [],
      "sourcePort": null,
      "targetPort": null,
    },
    "position": Point {
      "x": 0,
      "y": 0,
    },
    "type": "point",
  },
  "parent": LinkModel {
    "labels": Array [],
    "listeners": Object {},
    "options": Object {
      "id": "7",
      "locked": true,
    },
    "points": Array [
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "8",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "9",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
    ],
    "renderedPaths": Array [],
    "sourcePort": null,
    "targetPort": null,
  },
  "position": Point {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`PointModel All: Minimal config 1`] = `
PointModel {
  "listeners": Object {},
  "options": Object {
    "id": "1",
    "link": LinkModel {
      "labels": Array [],
      "listeners": Object {},
      "options": Object {
        "id": "1",
      },
      "points": Array [
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "2",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "3",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
      ],
      "renderedPaths": Array [],
      "sourcePort": null,
      "targetPort": null,
    },
    "position": Point {
      "x": 0,
      "y": 0,
    },
    "type": "point",
  },
  "parent": LinkModel {
    "labels": Array [],
    "listeners": Object {},
    "options": Object {
      "id": "1",
    },
    "points": Array [
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "2",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "3",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
    ],
    "renderedPaths": Array [],
    "sourcePort": null,
    "targetPort": null,
  },
  "position": Point {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`PointModel All: Selected 1`] = `
PointModel {
  "listeners": Object {},
  "options": Object {
    "id": "1",
    "link": LinkModel {
      "labels": Array [],
      "listeners": Object {},
      "options": Object {
        "id": "13",
      },
      "points": Array [
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "14",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
        PointModel {
          "listeners": Object {},
          "options": Object {
            "id": "15",
            "link": [Circular],
            "type": "point",
          },
          "parent": [Circular],
          "position": Point {
            "x": 0,
            "y": 0,
          },
        },
      ],
      "renderedPaths": Array [],
      "sourcePort": null,
      "targetPort": null,
    },
    "locked": true,
    "position": Point {
      "x": 0,
      "y": 0,
    },
    "selected": true,
    "type": "point",
  },
  "parent": LinkModel {
    "labels": Array [],
    "listeners": Object {},
    "options": Object {
      "id": "13",
    },
    "points": Array [
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "14",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
      PointModel {
        "listeners": Object {},
        "options": Object {
          "id": "15",
          "link": [Circular],
          "type": "point",
        },
        "parent": [Circular],
        "position": Point {
          "x": 0,
          "y": 0,
        },
      },
    ],
    "renderedPaths": Array [],
    "sourcePort": null,
    "targetPort": null,
  },
  "position": Point {
    "x": 0,
    "y": 0,
  },
}
`;
